packages/hash/frontend/src/blocks/page/Comments/CommentTextField.tsx

```diff
@@ -0,0 +1,213 @@
+import { createTextEditorView } from "../createTextEditorView";
+
+type CommentTextFieldProps = {
+  onClose: () => void;
+  onSubmit: (content: TextToken[]) => Promise<void>;
+};
+
+export const CommentTextField: FunctionComponent<CommentTextFieldProps> = ({
+  onClose,
+  onSubmit,
+}) => {
+  const viewRef = useRef<EditorView<Schema>>();
+  const [portals, renderPortal] = usePortals();
+  const { accountId } = useRouteAccountInfo();
+  const [loading, setLoading] = useState(false);
+  const editorContainerRef = useRef<HTMLDivElement>();
+
+  const updateEditorHandleKeyDown = useCallback(
+    (container: EditorView<Schema>, preventActions: boolean) => {
+      container.setProps({
+        handleKeyDown: (view, { shiftKey, key }) => {
+          if (!preventActions && !shiftKey) {
+            switch (key) {
+              case "Enter":
+                if (view.state.doc.content) {
+                  const { tokens } = textBlockNodeToEntityProperties(
+                    view.state.doc,
+                  );
+
+                  setLoading(true);
+                  onSubmit(tokens)
+                    .then(() => {
+                      onClose();
+                    })
+                    .finally(() => {
+                      setLoading(false);
+                    });
+                }
+
+                return true;
+
+              case "Escape":
+                onClose();
+                break;
+
+              default:
+                break;
+            }
+          }
+
+          return false;
+        },
+      });
+    },
+    [onClose, onSubmit],
+  );
+
+  const createEditor = useCallback(
+    (container: HTMLDivElement) => {
+      container.setAttribute("innerHTML", "");
+
+      const schema = createSchema({
+        doc: {
+          content: "inline*",
+        },
+        ...textTokenNodes,
+      });
+
+      const state = EditorState.create<Schema>({
+        schema,
+        plugins: [
+          keymap<Schema>(baseKeymap),
+          ...createFormatPlugins(renderPortal),
+          formatKeymap(schema),
+          createSuggester(renderPortal, accountId, container),
+          commentPlaceholderPlugin(renderPortal, "Leave a comment"),
+        ],
+      });
+
+      const view = createTextEditorView(
+        state,
+        container,
+        renderPortal,
+        accountId,
+      );
+
+      view.dom.classList.add(styles.Comment__TextField_Prosemirror_Input!);
+
+      updateEditorHandleKeyDown(view, false);
+      view.focus();
+
+      viewRef.current = view;
+    },
+    [accountId, renderPortal, updateEditorHandleKeyDown],
+  );
+
+  useEffect(() => {
+    if (viewRef.current) {
+      updateEditorHandleKeyDown(viewRef.current, loading);
+    }
+  }, [updateEditorHandleKeyDown, loading]);
+
+  useEffect(() => {
+    const editorContainer = editorContainerRef.current;
+    if (editorContainer) {
+      createEditor(editorContainer);
+    }
+  }, [editorContainerRef, createEditor]);
```

Comment:
This can be simplified a fair bit using some refs, the keymap plugin, and also ought to add an effect teardown. Let me know what you think? 

```suggestion
const editorContainerRef = useRef<HTMLDivElement>();

  const loadingRef = useRef(false);
  const eventsRef = useRef({ onClose, onSubmit });

  useLayoutEffect(() => {
    eventsRef.current = { onClose, onSubmit };
  });

  useEffect(() => {
    const editorContainer = editorContainerRef.current;
    if (editorContainer) {
      editorContainer.setAttribute("innerHTML", "");

      const schema = createSchema({
        doc: {
          content: "inline*",
        },
        ...textTokenNodes,
      });

      const state = EditorState.create<Schema>({
        schema,
        plugins: [
          keymap<Schema>({
            Enter(_, __, view) {
              if (!loadingRef.current && view?.state.doc.content) {
                const { tokens } = textBlockNodeToEntityProperties(
                  view.state.doc,
                );

                loadingRef.current = true;
                eventsRef.current
                  .onSubmit(tokens)
                  .then(() => {
                    eventsRef.current.onClose();
                  })
                  .finally(() => {
                    loadingRef.current = false;
                  });
                return true;
              }
              return false;
            },
            Escape() {
              if (!loadingRef.current) {
                eventsRef.current.onClose();
                return true;
              }
              return false;
            }
          }),
          keymap<Schema>(baseKeymap),
          ...createFormatPlugins(renderPortal),
          formatKeymap(schema),
          createSuggester(renderPortal, accountId, editorContainer),
          commentPlaceholderPlugin(renderPortal, "Leave a comment"),
        ],
      });

      const view = createTextEditorView(
        state,
        editorContainer,
        renderPortal,
        accountId,
      );

      view.dom.classList.add(styles.Comment__TextField_Prosemirror_Input!);
      view.focus();

      viewRef.current = view;

      return () => {
        view.destroy();
        viewRef.current = undefined;
      };
    }
  }, [accountId, renderPortal]);
```

Reply:
Thank you, made this change in bf080922247cd80e3653529dccf9938a7ef85877
Also had to make a change in `createFormatPlugins` to prevent an error when closing the comment input. I don't fully get what is happening there so please let me know if it breaks anything.